job/job.go

```diff
@@ -0,0 +1,308 @@
+	}
+	return &streamer{File: f}, nil
+}
+
+// NewJob creates a job with specified command but does not start it
+// NOTE: uuid.NewRandom() returns almost unique value and has very low chance of
+// collision. So, for simplicity, regenerating the UUID in case of
+// collision will be skipped.
+func NewJob(command string) *Job {
+	args := strings.Split(command, " ")
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil
+	}
+
+	job := &Job{
+		cmd:   exec.Command(args[0], args[1:]...),
+		UUID:  id,
+		Mutex: &sync.Mutex{},
+		doneC: make(chan struct{}),
+		status: JobStatus{
+			Status:   SCHEDULED,
+			ExitCode: 0,
+		},
+	}
+
+	return job
+}
+
+// Start starts the job. It will call Start() and Wait() on underlying
+// exec.Cmd. Any error will be reported to job's status
+func (j *Job) Start() error {
+	j.Lock()
+	// If job already started, return an error indicating so
+	if j.cmd.Process != nil {
+		return errors.New("job has already started")
+	}
+
+	// If job has already exited, then disallow rerun
+	// NOTE: user can always create a new job with same command
+	if j.status.Status == EXITED {
+		return errors.New("job has exited; cannot restart")
+	}
+	j.Unlock()
+
+	// If creating a file to store output fails, don't run the job
+	// TODO: this is not ideal solution. Ideally, we want to fall
+	// back to in-memory buffer for holding the output.
+	// and stream by reading the buffer.
+	// TODO: implement the this if have some time
+	outStreamer, err := NewStreamer()
+	if err != nil {
+		return errors.New("failed to create a file to stream output of the run")
+	}
+
+	j.outStreamer = outStreamer
+
+	// start() will call exec.Cmd.Start() which might return an error
+	// need to wait until result of Start() comes back so that we can
+	// immediately report it back to the caller
+	errC := make(chan error)
+	go j.start(errC)
+
+	// Channel will either return an error or will be closed and give nil error
+	// In any case, we need to return it back to the caller
+	err = <-errC
+
+	return err
+}
+
+// Status returns status of this job.
+// Users can determine whether specified job has completed or not via this function
+func (j *Job) Status() JobStatus {
+	j.Lock()
+	defer j.Unlock()
+	return j.status
+}
+
+// Stop terminates a running job. Returns immediately if job has already exited
+func (j *Job) Stop() error {
+	j.Lock()
+	if j.status.Status == EXITED {
+		return nil
+	}
+	j.Unlock()
+
+	pid, err := j.Pid()
+	if err != nil {
+		return err
+	}
+
+	pgid, err := syscall.Getpgid(pid)
+	if err != nil {
+		return err
+	}
+
+	err = syscall.Kill(-pgid, syscall.SIGTERM)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// Output returns a string channel which will be closed by library internally
+// once the job completes. Multiple calls to this function will create separate channels
+// so it is safe to stream output to multiple clients.
+func (j *Job) Output() chan []byte {
+	j.Lock()
+	defer j.Unlock()
+	rd, err := j.outStreamer.NewReader()
```

Comment:
This reader is never closed.

Reply:
startStream() should close it once it is done. will fix it next commit.